File: desktop/ai-app-bridge-cli/scripts/validation/notally/regression.py
"""Python independently interprets the same plan and calls the public executor."""
import json
import time
import traceback


def main(ctx):
    plan = ctx.inputs['plan']
    run_id = ctx.inputs['runId']
    steps = json.loads(json.dumps(plan['steps'], ensure_ascii=False).replace('@RUN@', run_id))
    now = lambda: int(time.time() * 1000)
    result = dict(language='python', planVersion=plan['version'], runId=run_id,
                  ok=False, startedAtMs=now(), steps=[], phases=[], actions=0, assertions=0)
    identity = None
    current_phase = None

    def save():
        with open(ctx.inputs['output'], 'w') as stream:
            json.dump(result, stream, ensure_ascii=False, indent=2)

    def call(args):
        reply = ctx.call('android-executor', {'feedback': 'off', **args})
        if not reply['ok'] or reply['result'].get('ok') is False:
            raise RuntimeError(str(reply))
        return reply['result']

    def observe(engine):
        return call({**identity, 'operation': 'observe', 'engine': engine})['observation']

    def matches(node, selector, observation):
        for key, value in selector.items():
            if key == 'within':
                parents = [n for n in observation['nodes'] if matches(n, value, observation)]
                if len(parents) != 1:
                    return False
                parent = node
                while parent and parent['nodeId'] != parents[0]['nodeId']:
                    parent = next((n for n in observation['nodes'] if n['nodeId'] == parent.get('parentId')), None)
                if parent is None:
                    return False
            elif key == 'id':
                expected = value if ':' in value else ctx.inputs['target']['packageName'] + ':id/' + value
                if node.get('resourceId') != expected:
                    return False
            elif key == 'classSuffix':
                if not (node.get('className') or '').endswith(value):
                    return False
            elif key == 'textContains':
                if not isinstance(node.get('text'), str) or value not in node['text']:
                    return False
            elif node.get(key) != value:
                return False
        return True

    def selected(observation, selector):
        return [n for n in observation['nodes'] if matches(n, selector, observation)]

    def act(observation, kind, node, fields, index):
        action = {'type': kind, **({'nodeId': node['nodeId']} if node is not None else {}), **fields}
        reply = call({**identity, 'operation': 'act', 'snapshotId': observation['snapshotId'],
                      'actionId': f'{run_id}-{index}-{result["actions"]}', 'action': action})
        result['actions'] += 1
        return reply['action']

    try:
        opened = call(dict(operation='open', serial=ctx.inputs['target']['serial'], packageName=ctx.inputs['target']['packageName'],
                           instrumentation=ctx.inputs['target']['packageName'] + '.test/androidx.test.runner.AndroidJUnitRunner',
                           testClass='io.github.mobileaidev.aiappbridge.generated.BridgeSessionTest',
                           activity='com.philkes.notallyx.presentation.activity.main.MainActivity',
                           leaseMs=1800000, timeoutMs=45000))
        identity = dict(serial=ctx.inputs['target']['serial'], packageName=ctx.inputs['target']['packageName'],
                        sessionId=opened['sessionId'], runtimeEpoch=opened['runtimeEpoch'])
        result['identity'] = identity
        result['flowStartedAtMs'] = now()
        for index, step in enumerate(steps):
            if step['phase'] != current_phase:
                current_phase = step['phase']
                result['phases'].append(dict(name=current_phase, startedAtMs=now(), firstStep=index))
                ctx.progress(dict(phase=current_phase, step=index, total=len(steps)))
            record = dict(index=index, phase=step['phase'], name=step['name'], operation=step['op'], startedAtMs=now())
            result['steps'].append(record)
            engine = step.get('engine', 'espresso')
            deadline = now() + step.get('timeoutMs', 7000)
            scrolls = 0
            while True:
                observation = observe(engine)
                found = selected(observation, step['selector']) if 'selector' in step else []
                if 'selector' not in step or len(found) == step.get('count', 1):
                    break
                if now() >= deadline:
                    result['failureObservation'] = observation
                    raise RuntimeError(f'Selector count {len(found)} at {index}: {step}')
                if step.get('seek') and scrolls < 10:
                    boxes = selected(observation, {'id': 'MainListView', 'visible': True})
                    assert len(boxes) == 1
                    act(observation, 'swipeDown', boxes[0], {}, index)
                    scrolls += 1
                else:
                    time.sleep(0.1)
            if step['op'] == 'assert':
                result['assertions'] += 1
                record['actual'] = found
            elif step['op'] == 'action':
                node = found[0] if found else None
                if step.get('scroll') is True and node is not None and not node['visible']:
                    act(observation, 'scrollTo', node, {}, index)
                    observation = observe(engine)
                    found = selected(observation, step['selector'])
                    if len(found) != 1 or not found[0]['visible']:
                        raise RuntimeError('Scroll did not reveal exact target')
                    node = found[0]
                record['mechanism'] = act(observation, step['action'], node, step.get('fields', {}), index)
            else:
                raise RuntimeError('Unknown plan operation ' + step['op'])
            record['passed'] = True
            record['elapsedMs'] = now() - record['startedAtMs']
            save()
        result['flowElapsedMs'] = now() - result['flowStartedAtMs']
        result['finalObservation'] = observe('espresso')
        result['ok'] = True
    except Exception:
        result['error'] = traceback.format_exc()
        if identity is not None and 'failureObservation' not in result:
            try:
                result['failureObservation'] = observe('espresso')
            except Exception as error:
                result['observationError'] = str(error)
    finally:
        if identity is not None:
            try:
                result['close'] = call({**identity, 'operation': 'close', 'timeoutMs': 30000})
            except Exception as error:
                result['ok'] = False
                result['closeError'] = str(error)
        result['elapsedMs'] = now() - result['startedAtMs']
        save()
    return result

File: desktop/ai-app-bridge-cli/scripts/validation/notally/regression.js
'use strict';

// Each UI action goes through the published executor. This is a real JS Script,
// not a wrapper that launches a fixed test method and calls that Python support.
module.exports.main = async function main(ctx) {
  const fs = require('node:fs');
  const { plan, output, runId } = ctx.inputs;
  const steps = JSON.parse(JSON.stringify(plan.steps).replaceAll('@RUN@', runId));
  const result = { language: 'javascript', planVersion: plan.version, runId,
    ok: false, startedAtMs: Date.now(), steps: [], phases: [], actions: 0, assertions: 0 };
  let identity, currentPhase;
  const save = () => fs.writeFileSync(output, JSON.stringify(result, null, 2));
  const call = async args => {
    const r = await ctx.call('android-executor', { feedback: 'off', ...args });
    if (!r.ok || r.result.ok === false) throw new Error(JSON.stringify(r));
    return r.result;
  };
  const observe = async engine => (await call({ ...identity, operation: 'observe', engine })).observation;
  function matches(node, selector, observation) {
    for (const [key, value] of Object.entries(selector)) {
      if (key === 'within') {
        const parents = observation.nodes.filter(n => matches(n, value, observation));
        if (parents.length !== 1) return false;
        let p = node;
        while (p && p.nodeId !== parents[0].nodeId) p = observation.nodes.find(n => n.nodeId === p.parentId);
        if (!p) return false;
      } else if (key === 'id') {
        if (node.resourceId !== (value.includes(':') ? value : ctx.inputs.target.packageName + ':id/' + value)) return false;
      } else if (key === 'classSuffix') {
        if (!node.className?.endsWith(value)) return false;
      } else if (key === 'textContains') {
        if (typeof node.text !== 'string' || !node.text.includes(value)) return false;
      } else if (node[key] !== value) return false;
    }
    return true;
  }
  function selected(o, selector) { return o.nodes.filter(n => matches(n, selector, o)); }
  async function act(o, type, node, fields, index) {
    const r = await call({ ...identity, operation: 'act', snapshotId: o.snapshotId,
      actionId: runId + '-' + index + '-' + result.actions,
      action: { type, ...(node ? { nodeId: node.nodeId } : {}), ...fields } });
    result.actions++;
    return r.action;
  }
  try {
    const opened = await call({ operation: 'open', serial: ctx.inputs.target.serial,
      packageName: ctx.inputs.target.packageName,
      instrumentation: ctx.inputs.target.packageName + '.test/androidx.test.runner.AndroidJUnitRunner',
      testClass: 'io.github.mobileaidev.aiappbridge.generated.BridgeSessionTest',
      activity: 'com.philkes.notallyx.presentation.activity.main.MainActivity',
      leaseMs: 1800000, timeoutMs: 45000 });
    identity = { serial: ctx.inputs.target.serial, packageName: ctx.inputs.target.packageName,
      sessionId: opened.sessionId, runtimeEpoch: opened.runtimeEpoch };
    result.identity = identity;
    result.flowStartedAtMs = Date.now();
    for (const [index, step] of steps.entries()) {
      if (step.phase !== currentPhase) {
        currentPhase = step.phase;
        result.phases.push({ name: currentPhase, startedAtMs: Date.now(), firstStep: index });
        await ctx.progress({ phase: currentPhase, step: index, total: steps.length });
      }
      const record = { index, phase: step.phase, name: step.name, operation: step.op, startedAtMs: Date.now() };
      result.steps.push(record);
      const engine = step.engine || 'espresso';
      const deadline = Date.now() + (step.timeoutMs || 7000);
      let o, found, scrolls = 0;
      do {
        o = await observe(engine);
        found = step.selector ? selected(o, step.selector) : [];
        const desired = step.count === undefined ? 1 : step.count;
        if (!step.selector || found.length === desired) break;
        if (step.seek && scrolls++ < 10) { const boxes = selected(o, { id: 'MainListView', visible: true }); if (boxes.length !== 1) throw Error('Scroll container not unique'); await act(o, 'swipeDown', boxes[0], {}, index); } else await new Promise(resolve => setTimeout(resolve, 100));
      } while (Date.now() < deadline);
      if (step.selector && found.length !== (step.count === undefined ? 1 : step.count)) {
        result.failureObservation = o;
        throw new Error('Selector count ' + found.length + ' at ' + index + ': ' + JSON.stringify(step));
      }
      if (step.op === 'assert') {
        result.assertions++;
        record.actual = found;
      } else if (step.op === 'action') {
        let node = found[0];
        if (step.scroll === true && node && !node.visible) {
          await act(o, 'scrollTo', node, {}, index);
          o = await observe(engine);
          found = selected(o, step.selector);
          if (found.length !== 1 || !found[0].visible) throw new Error('Scroll did not reveal exact target');
          node = found[0];
        }
        record.mechanism = await act(o, step.action, node, step.fields || {}, index);
      } else throw new Error('Unknown plan operation ' + step.op);
      record.passed = true;
      record.elapsedMs = Date.now() - record.startedAtMs;
      save();
    }
    result.flowElapsedMs = Date.now() - result.flowStartedAtMs;
    result.finalObservation = await observe('espresso');
    result.ok = true;
  } catch (error) {
    result.error = String(error.stack || error);
    if (identity && !result.failureObservation) {
      try { result.failureObservation = await observe('espresso'); }
      catch (observationError) { result.observationError = String(observationError); }
    }
  } finally {
    if (identity) {
      try { result.close = await call({ ...identity, operation: 'close', timeoutMs: 30000 }); }
      catch (error) { result.ok = false; result.closeError = String(error); }
    }
    result.elapsedMs = Date.now() - result.startedAtMs;
    save();
  }
  return result;
};
